16.0 mig payment token uniquify

Honeyxilia opened the pull request.

fcayre committed:
Create the payment_token_uniquify module and a first documentation

fcayre committed:
Implement payment token obsolescence logic

When a new payment token is created for a partner, previous tokens
that are present below the first parent of this partner (including
itself) that is marked with the uniquify_token field set to true
will be considered obsolete.

fcayre committed:
Add uniquify_token in the partner form view

fcayre committed:
Add a sequence field to the obsolete token actions to enforce execution order

fcayre committed:
Add debugging to show action execution order in the logs

fcayre committed:
Reverse the logic of the partner field used to define the obsolescent tokens

... and update the doc.

fcayre committed:
Obsolete tokens deactivate action also clears their partner's payment_token_id field

Otherwise the inactivated token are still accessible through the partner, which is weird.

-------------

github-grap-bot committed:
payment_token_uniquify 12.0.1.0.1

fcayre committed:
Reset payment date preferences of partners which payment token is reset

This way, if such a partner sign a new mandate in the future, the
payment date will be correctly set instead of preserving an old,
insignificant one.

github-grap-bot committed:
payment_token_uniquify 12.0.1.0.2

github-grap-bot committed:
[UPD] README.rst

fcayre committed:
Small test refactorings to help understanding & debugging

fcayre committed:
Remove the isolated_payment_token partner field

It is useless now that we use the company as a barrier to determine which payment tokens
are obsolete after a new one has been signed. This is much more consistant with other native
odoo mecanisms like address_get or the like.

github-grap-bot committed:
[BOT] post-merge updates

fcayre committed:
Use __name__ as logger names, not __file__

This makes it easier to mute them in tests.
This is the standard Odoo way to name them and standardize it in our
code base.

github-grap-bot committed:
[BOT] post-merge updates

Honeyxilia committed:
[IMP] payment_token_uniquify: black, isort, prettier

Honeyxilia committed:
[MIG] payment_token_uniquify: Migration to 16.0

Honeyxilia committed:
Removed depreciated @api.returns decorator

Honeyxilia committed:
Changed elements related to payment.acquirer to payment.provider migration

Honeyxilia committed:
Updated README

Honeyxilia committed:
Added contract_payment_auto dependency

Some operations on `res.partner` are reliant on having a specific payment token associated with a partner. This behavior was added through the contract_payment_auto module, so we inherit it explicitly in this module.

Honeyxilia committed:
Added batch create support for payment.token

Honeyxilia committed:
[test] Removed deleted 'name' field from 'payment.token' creation

Honeyxilia committed:
[test] Replaced setUp method by setUpClass

Honeyxilia committed:
[test] Replaced SavepointCase by TransactionCase

Honeyxilia committed:
[test] Removed useless operations

In the new_payment_token method, if the set_as_partner_token is not set to False, then the newly created token is assigned to the given partner already.

Honeyxilia committed:
[test] Added new token creation to improve test coverage

This operation allows for several operations to be tested :
- new_payment_token with set_as_partner_token set to False
- The obsolescence of tokens without a partner's assigned token

Honeyxilia committed:
[test] Removed setUp method and added deactivate operation

Since there is only one test method in this class, I moved the declarations in the test method itself.
This also allows for a first deactivate operation, to cover the case where there are no token to render obsolete.

fcayre commented on payment_token_uniquify/__manifest__.py:
```diff
@@ -11,7 +11,7 @@
     "license": "AGPL-3",
     "depends": [
-        "payment",
+        "account_payment",
```
Ici plutôt que de modifier les dépendances il vaut mieux modifier la vue pour ne pas dépendre du champ journal_id.

Honeyxilia replied:
Fait

fcayre commented on payment_token_uniquify/tests/common.py:
```diff
@@ -61,7 +61,7 @@ def _trigger_obsolescence(self, *action_refs, **new_partner_kwargs):
         for action_ref in action_refs:
             if "." not in action_ref:
-                action_ref = "commown.obsolescence_action_" + action_ref
+                action_ref = "payment_token_uniquify.obsolescence_action_" + action_ref
```
D'après ce que je vois dans le code de la v12, cette "feature" est utilisée par le module commown.
Il vaudrait mieux ici forcer l'obligation de mettre une référence complète (avec un "."), ce qui conduira à un crash dans les tests du module commown, que l'on modifiera en conséquence.

Honeyxilia replied:
Fait - j'ai ajouté un check assertIn, avec un message d'indication

Honeyxilia committed:
[test] Removed action ref. autocomplete in _trigger_obsolescence

Beforehand, to allow shorter references to actions in tests, mainly in the commown module, which then auto-completes with the start of reference for an action declared in the commown module.
However, this makes this part out of place in the payment_token_uniquify module, reducing its readability.

So, we removed it for visibility

Honeyxilia committed:
Modified view to not depend on moved field

Previously, we used the `journal_id` field of `payment.acquirer` to concatenate the obsolescence token field right after it.
However, the `journal_id` field has been migrated from the `payment` module to the `account_payment` module, meaning we can't use it without changing the dependency (and adding more modules to download), we instead directly inherit the original view to add the obsolescence tokens field at the same place.
Also, since the payment_followup group is invisible by default, we add an attribute change to invisible to insure it'll be visible.

NOTE: since there are no menuitems linked with actions in payment to access payment.provider records, installing account_payment locally is still required to check if the field is correctly added or not, because it's where the menuitem that allows frontend (browser) access to the provider records.

fcayre commented on payment_token_uniquify/tests/common.py:
```diff
@@ -60,8 +60,12 @@ def _trigger_obsolescence(self, *action_refs, **new_partner_kwargs):
+                action_ref,
+                "_trigger_obsolescence: Please use the full reference of the "
+                "action reference id. (<full module name>.<action record id>).",
+            )
```
Pour moi cet assert est vraiment overkill : si tu l'enlèves complètement le code est plus court (donc moins effrayant) et le `self.env.ref`va planter en disant qu'il manque un ".", ça suffit largement (KISS).

fcayre commented on payment_token_uniquify/tests/test_payment_token.py:
```diff
@@ -2,22 +2,21 @@
-        super().setUp()
+    def test_action_deactivate(self):
+        "Obsolete tokens must become inactive when this action is used"
+        # Configure provider, and trigger obsolescence when no tokens are set
```
Peut-être ajouter en commentaire pourquoi tu fais ça ici :  pour améliorer la couverture (pas important cependant).

9184Lucky commented:
/ocabot rebase

9184Lucky commented:
/ocabot merge patch
